File: 101_Symmetric_Tree/xsy/solution.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

    def __repr__(self):
        return str(self.val)

class Solution:
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        for i, j in zip(self.preOrder(root.left), self.mirrorPreOrder(root.right)):
            if i != j:
                return False
        return True

    def preOrder(self ,root):
        """中左右遍历"""
        if not root:
            yield None
        else:
            yield root.val
            yield from self.preOrder(root.left)
            yield from self.preOrder(root.right)

    def mirrorPreOrder(self, root):
        """中右左遍历"""
        if not root:
            yield None
        else:
            yield root.val
            yield from self.mirrorPreOrder(root.right)
            yield from self.mirrorPreOrder(root.left)

File: 101_Symmetric_Tree/xsy/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSym(l, r *TreeNode) bool {
	if l != nil and r != nil and l.Val == r.Val {
		return isSym(l.Left, r.Right) && isSym(l.Right, r.Left)
	} else {
		return l == r
	}
}

func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}    
	return isSym(root.Left, root.Right)
}
